wip: implement Rename transform

Omar-Elrefaei opened the pull request:
This should be fully functional, but I think the usage of `push!!` will need to be redesigned to be more functional and efficient like other transforms. That is the place where I think I need a little help: I am not sure how to make it so that I don't need `push!!` and use `𝒯 = (; zip(anames, acols)...)` instead like the rest of the of the transforms.

juliohm commented on src/TableTransforms.jl:
```diff
@@ -7,7 +7,7 @@ module TableTransforms
 using ScientificTypes
 using Distributions: Normal
-using Transducers: tcollect
+using Transducers: tcollect, push!!
```
We don't need `push!!` to implement `Rename`.

juliohm commented on src/TableTransforms.jl:
```diff
@@ -29,9 +29,10 @@ export
   Select,
   Reject,
-  Identity,
+  Rename3,
```
This export is not correct. Also keep `Identity` in the line it was placed.

Omar-Elrefaei replied:
> Also keep Identity in the line it was placed.
I was making it inline with the order of includes in `transforms.jl`

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,43 @@
+# ------------------------------------------------------------------
+
+""" 
+    Rename(Dict(:col₁ => :newcol₁, :col₂ => :newcol₂, ..., :col₁ => :newcolₙ))
```
This is an internal signature. After we make it work with `Dict` we will only document the VarArg version to end users. So this string should read `Rename(:col1 => :newcol1, :col2 => :newcol2, ...)`.

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,43 @@
+
+Tha transform that renames `col₁` to `newcol₁`, `col₂` to `newcol₂`, ...
+""" 
+struct Rename3 <: Stateless
```
We don't want `Rename3`.

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,43 @@
+end
+
+function apply(transform::Rename3, table)
+    new_table = (;)
```
Before we start applying the transform, we need to make sure that the input table has the columns specified in `Rename`.

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,43 @@
+function apply(transform::Rename3, table)
+    new_table = (;)
+
+    for col_name in Tables.columnnames(table)
+        col_value = Tables.getcolumn(table, col_name)
```
Try to follow the code style in the other transforms. We avoid variable names with underscore for example. We use 2 whitespaces for indentation, etc.

juliohm reviewed (changes requested):
Attached a first round of reviews.

Omar-Elrefaei commented:
Ooof, my bad for half of these things. I guess that what you get from sending a PR of after-midnight code without a review. I will fix the formatting and naming right away, but that I am not 100% sure about is how to phase out `push!!` ?

Omar-Elrefaei commented:
Is it as simple as looping through the results of `Tables.columnnames(table)`, replacing the names that need to be renamed, and then `(; zip(names, cols)...)` ??

juliohm commented:
Exactly, you can write it in a single:

```julia
newcols = map(cols) do col
  # do the renaming
end
```

See the other transforms for examples.

Omar-Elrefaei commented:
I'm sorry, I screwed up git a bit. I'll comment when it's ready for review.

Omar-Elrefaei commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,37 @@
+function apply(transform::Rename, table)
+  oldnames = Tables.columnnames(table)
+  newnames = map(oldnames) do oldname
+    if oldname in keys(transform.names) 
+      transform.names[oldname]
+    else
+      oldname
+    end
```
This could be a simple:
```(oldname in keys(transform.names))  ? transform.names[oldname] : oldname```
or even:
`if (oldname in keys(transform.names)) transform.names[oldname] else oldname end`

juliohm replied:
The ternary operator is nice, specially if you save a new variable `names = transform.names` before the map to shorten it up.

Omar-Elrefaei commented:
I believe its ready for another review.

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,37 @@
+  names::Dict{Symbol,Symbol}
+end
+
+function apply(transform::Rename, table)
```
I think we need to `@assert` to make sure that the dictionary doesn't contain invalid oldnames. That is, before the map, we should make sure that the `keys(transform.names) \subseteq oldnames`.

Omar-Elrefaei replied:
I can add an assert if you want (maybe it's a UX issue?), but right now due to the `if oldname in keys(transform.names)` invalid oldnames are gracefully ignored.

juliohm replied:
Yes, we want to assert to alert end users that they specified incorrect column names 👍🏽

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,37 @@
+      oldname
+    end
+  end
+  acols = [i for i in Tables.columns(table)]
```
I think this is not fully compatible with the Tables.jl interface. The result of `Tables.column` is not necessarily iterable. We need to use `Tables.getcolumn(cols, name)` to access a specific column.

Omar-Elrefaei replied:
TIL. I noticed that it was not working with DataFrames, but it was also not working with JuliaDB, and I saw other `Tables.column` instances throughout the codebase so I thought it was fine for now. Turns out JuliaDB does not fully implement Tables.jl, and your other usages were not iterating.

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,37 @@
+function revert(transform::Rename, table)
+  # reversing the key-value pairs of the Dict
+  new_names = Dict()
+  for (new, old) in transform.names
```
I think you you misplaced `new` and `old` in this loop. Can you please add tests to the transform so that we catch basic bugs?

Omar-Elrefaei replied:
I will add tests (might take a little of poking around given this is my first forte into package testing.)
But I have been testing continuously while developing (I don't feel comfortable while writing code otherwise) and it works perfectly. Both reverting and appling invalid oldnames are handled as expected.
I can see why you said they might be misplaced, semantically I should probably swap them both in the for loop and the dict assignment (line32).

juliohm replied:
Are you familiar with Julia's testing infrastructure? All you need to do is activate the folder of the package and type `] test`. You then add more tests and rerun.

Omar-Elrefaei replied:
> Are you familiar with Julia's testing infrastructure?

No

> All you need to do is activate the folder of the package and type `] test`

that sounds like a simple workflow. I have used the package manager a lot but as a user, never as a package dev.
I can see the simple format in `test/transforms.jl` and I can add to that. 
Question: is there a way to just run the tests I'm developing instead of everything in there?

juliohm replied:
Never used but I think Retest.jl has this functionality: https://github.com/JuliaTesting/ReTest.jl

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,37 @@
+    new_names[old] = new
+  end
+
+  reversed_transform = Rename(new_names)
+  apply(reversed_transform, table) |> first
```
We don't want create a new transform to revert. What we need to do is save the old names in the cache and use the cache to revert the operation.

juliohm replied:
Alternatively you can refactor this common set of instructions into an separate function `_rename` that does the actual work, and use this function in multiple places. This will facilitate future rearrangements of the code, and is good software engineering practice.

Omar-Elrefaei replied:
I am gonna do it this second way, because I am not sure if there is anything useful to put in that cache.

So it is ok to: 
- refactor that `_rename`, 
- call it twice instead of creating another transform, 
- and not use `cache`. (pass it around as `nothing` instead)

juliohm reviewed (changes requested):
Review attached.

juliohm commented:
@Omar-Elrefaei did you have time to update this PR?

Omar-Elrefaei commented:
Unfortunately no, I haven't been getting a chance to continue the work.



On Thu, Nov 18, 2021, 9:06 AM Júlio Hoffimann ***@***.***>
wrote:

> @Omar-Elrefaei <https://github.com/Omar-Elrefaei> did you have time to
> update this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaML/TableTransforms.jl/pull/15#issuecomment-972895773>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIXXL6RMMX5QOWLZOVZCS3UMUB5NANCNFSM5IACTAYQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

juliohm commented:
@Omar-Elrefaei you mean that this PR should be closed?

Omar-Elrefaei commented:
> @Omar-Elrefaei you mean that this PR should be closed?

No, if not today after my dayjob, I'll definitely be sending updates in the weekend.

Omar-Elrefaei commented:
I think it's ready for another review.
I have:
- [x] added the assert
- [x] added tests
- [x] refactored `_rename` out
- [x] made the varargs constructor

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,50 @@
+
+function revert(transform::Rename, table, cache)
+  # reversing the key-value pairs of the Dict
+  newnames = Dict()
+  for (old, new) in transform.names
+    newnames[new] = old
+  end
```
```suggestion
  newnames = Dict(new => old for (old, new) in transform.names)
```

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,47 @@
+  oldnames = Tables.columnnames(table)
+
+  # check if requested renames exist in the table
+  diff = setdiff(keys(names), oldnames) .|> String |> Tuple
+  @assert length(diff) == 0
+    "The following column[s] were not found in the source table: $diff"
```
```suggestion
  @assert keys(names) ⊆ oldnames "invalid column names"
```

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,45 @@
+  # check if requested renames exist in the table
+  @assert keys(names) ⊆ oldnames "invalid column names"
+
+  newnames = map(oldnames) do oldname
```
```suggestion
  # use new names if necessary
  newnames = map(oldnames) do oldname
```

juliohm commented on src/transforms/rename.jl:
```diff
@@ -0,0 +1,46 @@
+# ------------------------------------------------------------------
+
+"""
+  Rename(:col₁ => :newcol₁, :col₂ => :newcol₂, ..., :col₁ => :newcolₙ)
```
```suggestion
    Rename(:col₁ => :newcol₁, :col₂ => :newcol₂, ..., :col₁ => :newcolₙ)
```

juliohm commented:
Awesome improvements @Omar-Elrefaei , I am merging it.

The test that is failing is unrelated. For some reason one of our visual tests is failing randomly. Do you think you can take a look at this issue? Maybe we are missing some `Random.seed!` call somewhere?

Omar-Elrefaei commented:
Great! 
Thank you very much @juliohm for the patient and valuable feedback.

I'll open an issue so that not to forget to investigate the failing tests. 
Looking forward to my next PR 😄
